[GRID 7] Dynamic image loader

Koerty opened the pull request:
This is a pretty big task, so I created this PR before it's finished so you can take a first look and give me some feedback if you want to. It is not completely stable, as it still have some bugs or unintended behaviors (list at the end of this message), but the base mechanics are already here and working as intended.

I rewrote the dynamic image loader from scratch, trying to make a simpler version than the original one, sadly, it isn't possible, you will understand why very soon.

## Current behavior
The dynamic image loader aim to load just the textures displayed on the screen + 1 extra row on top and on the bottom, so the user is less likely to see the magic happening.

For this, I used 2 var : **target min range** and **target max range**, those are updated every time the grid scrolled up or down (which mean games displayed on screen changed).

This happen "asynchronously", so it don't freeze when loading big amount of images, only one image can be loaded and one can be unloaded by frame (call to the update functions).

For this, 2 other var are used : **current min range** and **current max range**. Once every frame, both the **current min and max range** go one step forward to **target min and max range**, loading or unloading images on their way.

This work pretty good, even if the user scroll down the grid at the highest speed, but it doesn't fit all use case. Sometimes, the range change a lot, for example when the user select a random game and have a big game collection so the current range and the next target range have no image in common.

For this I added a 2nd mod, badly named "emergency mod", when it happen, on the next frame all the currently loaded textures will be unloaded all at the same time, then the **current min and max range** will both be set to **target min range**. At this moment the "emergency mod" is exited and the dynamic image loader will get back to normal behavior, this mean, since the **current max range** is set to **target min range**, it will go back, step by step, to **target max range**, loading the images on its way.

## Known bugs or unintended behaviors
- When the menu is opened by pressing "select", all the images are reseted to default, even if they are not unloaded
- When switching from one system to another, the current range isn't unloaded

## Testing
I left some debug print for now, as it help understanding what's going on.

Also, right now the system try to keep all the displayed images on screen loaded + 1 extra row and top and bottom, this can be changed to "-1 extra row" to better see what's happening, by changing : 
Line 331 : 
`nextTargetRange.min = startPos - mGridDimension.x();`
to
`nextTargetRange.min = startPos + mGridDimension.x();`

Line 336 : 
`nextTargetRange.max = startPos + (mGridDimension.x() + 1) * mGridDimension.y();`
to
`nextTargetRange.max = startPos + (mGridDimension.x() - 1) * mGridDimension.y();`

**Note that doing this may introduce some bugs, but you can better see base behavior.**

Koerty commented:
@jrassa This look good to me. I still need to run more tests on my pie and maybe do some cleaning/refactoring if I find some, but it work as intended and got no bugs so far.

Koerty commented:
On my laptop there is zero issue, but on my pie (Pie 3 model B) which have less processing power, the dynamic image loader can't follow at highest scroll speed.

Maybe we should consider diminishing the scrolling speed of the grid, as we already scroll 3 or 4 times faster than in the TextList, since the grid usually have 3 or 4 columns.

I'm not sure this will completely solve the problem, but would greatly diminish it impact.

Another solution would be to complexify the dynamic loader code, by loading more images in the direction the user is scrolling (like : load what images are on screen + 3 rows in user scrolling direction + 1 row in the other direction).

What's your thought on this @jrassa ?

hex007 commented:
I would like to suggest another method. Just like how videos are not immediately shown while fast scrolling, we can limit the images to load after some delay once the user stops scrolling. Does that sound plausible ?

Koerty commented:
@hex007 I think this is a slightly different issue, as the user mostly use the images to know what game he is looking at in the grid view, while in video view the video is just an extra, as he use the text list to select his game.

Koerty commented:
@hex007 If we do that maybe we should consider add an overlay like steam big picture do : when the user is scrolling at highest speed, the first letter of the current game name is wrote in big in the middle of the screen, and the rest of the screen is darkened.

hex007 commented:
Any other devs care to pitch their ideas?

pjft commented:
@jrassa is correct on both accounts. There is already support for the high speed scrolling layer with letters in the normal views, and I agree they could be used here. 

I believe the TextureResource class (name may be inaccurate, on my phone) handles image loading and unloading to keep things within the allowed GPU memory.

What you may want to do is, similar to the carousel view, to cache the non-visible neighboring images so that they're already loaded when scrolling.

I believe that logic lives in SystemView. Once again, apologies for not linking any code but can look later. 

What exactly were you aiming for in this new dynamic loader?

Best.

pjft commented:
On phone, can't edit previous post.

I think you're doing exactly what I was suggesting, the only thing that's not clear from the doc's you wrote and that may help is that when scrolling in a direction you may want to cache more lines ahead and ignore the ones before. 

If that's already done, then please ignore. :)

pjft commented:
Here:

https://github.com/RetroPie/EmulationStation/pull/129/files

pjft commented:
See the logic in SystemView renderExtras or renderCarousel for how that's being handled. It seems that you won't need a large portion of the code you have, but rather simply ask it to render those images. 

You shouldn't need to handle any texture loading and unloading.

Koerty commented:
Damn, I didn't know that, would have been nice to give me that information before I started all this stuff, that would have saved me a lot of time.

I will take a look at how it work once I'm done with the NinePatchComponent upgrade.

Koerty commented:
So, if I understood correctly, I just need to call the static "get" function of the class TextureResource, with its last parameter, "dynamic" to true, and it will instantly load the texture, and unload them later on when the vram limit is reached ?

pjft commented:
I imagine that is what will eventually happen but I don't think you need to call it yourself, they should all be abstracted from you. 

You should do exactly what SystemView does and call render for those elements. You should not have to call TextureResource yourself, though I may be misremembering. It's been a while since I looked into that code myself.

Koerty commented:
I think I got it. I need to modify the ImageGridData to only hold the texture path : 
```
struct ImageGridData
{
	std::string texturePath;
};
```

Instead of a TextureResource like it was done until now : 
```
struct ImageGridData
{
	std::shared_ptr<TextureResource> texture;
};
```

Then this texture path will be set to the ImageComponents of the GridTileComponents, so the ImageComponents with the help of TextureResource and TextureDataManager will automatically decide what texture to load and unload.

That sound good to me, and way easier to manage, but ... that mean 98% of the code added by this PR is completely useless ... damn ...

Koerty commented:
Wait no, that wouldn't work, to load textures ahead I would also need to create extra GridTileComponents positioned offscreen.

pjft commented:
Well, tell me something. How does the grid work - are the elements static and you just change the images they load, or do you create all the elements upfront and then move the camera?

What happens, high level, when you scroll one row?

Koerty commented:
We have a fixed amount of GridTileComponent, they never move, and correspond to the grid dimensions. For example for a 3x4 grid we have 12 GridTileComponent.

The ImageGridComponent, is an IList of ImageGridData, ImageGridData being the structure where we store the textures.

When the grid scroll, the ImageGridComponent take the 12 textures corresponding to the current position, and set them to the ImageComponent of the GridTileComponent.

Koerty commented:
So, if we want to load more textures to plan ahead, for example 3 rows in the scrolling direction, that mean we have to create 6 extra rows of GridTileComponent (3 for each side) **and** render them. That mean we either have to position them offscreen, or add a fake background on top of them, which isn't possible at the moment since the grid tiles don't support zIndex and are always rendered on top of everything.

pjft commented:
I see. Indeed, with SystemView this works as we render the entire carousel and then the camera does the rest.

In this case, we're not going to load all images for all grids all at once, so it'll be different. 

I don't think we'll need to create extra GridTileComponents and render them, as that'd be a bit of a hack in this specific architecture. 

I think a lot of the logic you put in place so far can and will be used here. I just feel that the loadTexture method is somewhat redundant with what the TextureResource class already does - I suspect you could replace that entire method with just `mEntries[index].data.texture = TextureResource::get(imagePath, false, true);` or close to that, as TextureResource::get should in theory do that.

https://github.com/Koerty/EmulationStation/blob/6a3e55a9591dfebd234ad04618dd7cb930d32b03/es-core/src/resources/TextureDataManager.cpp#L49

Now, the question is how to cache those upfront. You can certainly keep in memory a lookahead array of +3 and -3 textures and load them upfront, and see if that helps. It's just that the overall code can be simplified so you're not doing things that should not be done by the view.

jrassa commented:
> So, if we want to load more textures to plan ahead, for example 3 rows in the scrolling direction, that mean we have to create 6 extra rows of GridTileComponent (3 for each side) and render them. That mean we either have to position them offscreen,

You wouldn't necessarily have to position them offscreen.  The way that the System View works is it renders them in their properly relative position (which may or may not be off screen), but relies on the a clipRect defined to the size/position of the carousel to ensure they are not actually visually rendered.  This approach would also have the potential advantage of allowing grid tiles to slide on scroll vs. just changing instantly, which I believe you expressed a desire in implementing at some point.

This may be the best approach ultimately, but I understand that this represents a big change from how things work now.  I agree with @pjft that most of the logic you have in place is good and the loading can probably be replaced with using `TextureResource::get` and the unloading code won't be needed since the TextureDataManager should handle it for you.

> or add a fake background on top of them which isn't possible at the moment since the grid tiles don't support zIndex and are always rendered on top of everything.

Not that I would recommend this approach, but for clarification, since the grid tiles are rendered by the ImageGrid, I believe that they effectively inherit the zIndex of the grid, which is set to 20.  zIndex values really only apply to the immediate children of the views.  They determine the order in which the views children are rendered by the view.

Koerty commented:
> relies on the a clipRect defined to the size/position of the carousel to ensure they are not actually visually rendered. This approach would also have the potential advantage of allowing grid tiles to slide on scroll vs. just changing instantly, which I believe you expressed a desire in implementing at some point.

How did I missed that, I was looking for something like this "clipRect" since I started to work on the grid view.

Yes I would like to implement scrolling tiles, that would complexify a bit the code but I feel like it will greatly enhance user experience. Right now, when you scroll left or right you can actually see the selector moving, but when you scroll down or up and you are not close to the end or beginning, the selector don't move, just the images move, so it feel like ... weird. It feel like the grid didn't scroll at all.

>  I agree with @pjft that most of the logic you have in place is good and the loading can probably be replaced with using TextureResource::get and the unloading code won't be needed since the TextureDataManager should handle it for you.

Since texture management can already be handled by the TextureDataManager, that mean all this 200 lines of code have only one purpose left : load 1 texture per frame. This look completely overkill to me.

I think I'm going to add 3 extra row on both side as I said (where 2~4 of them will be empty), and leave all the texture management stuff to the TextureDataManager.

This way I just have to add the code to load 1, 2 or 3 extra row in front, depending on scrolling speed, and 1 row behind. And also change a bit the code of the render function and the one that create and position tiles.

That way the code will be smaller, cleaner, and already half ready for the scrolling tile feature.

Koerty commented:
I was doing some testing and I want to share with you my results.

This test were made on my laptop, I tried to scroll down in the grid from the first game to the last game, then from the last game to the first game. This make me only reach the 2nd highest scrolling speed (my game collection isn't big enough for the highest).

I started from RetroPie's EmulationStation master branch, and just replaced `std::shared_ptr<TextureResource> texture;` in the `ImageGridData` struct by an `std::string texture;`.

Without that changes, the framerate show a consistent 1000~1100 fps, the "Tex VRAM" is 72.53, the "Tex Max" is 1309.46

After this changes, when scrolling down the fps can drop to 800, the "Tex VRAM" is at 15.66, "Tex Max" is at 16.06, but when scroll up, the fps can drop below 1.

I think this is because the TextureDataManager is unloading textures he is going to reload right after. When scrolling down in a 3x4 grid, the image set to tile number 9 will now be set to tile number 6, this mean when we set the image to tile number 6, it's still in memory (as we didn't reach tile 9 yet), so it don't have to load it again. But when scrolling up, the image of the tile number 6 is updated, which mean the old TextureResource is unloaded, and then reloaded when we set the image of the tile 9.

All of this to say that whatever choice we make, we have to carefully think about it and carefully try them out. Also their is a lot of optimizations to be done on the grid, and I'm ensure when and how address them all yet. For example, right now, each time the cursor position changed, we reset all tiles image, all tiles to selected or unselected, all tiles to visible or not. We could do more complex operations, like save the last cursor position, only changes the texture and visibility if the grid scrolled (if cursor position changed row), only set the tile at last cursor position to unselected and the tile to new cursor position to selected so we don't call heavy operations like resize the image if it didn't changed.

pjft commented:
Thanks for the tests, and I do agree that these need to be tested out thoroughly.

I am not sure that the TextureResource would be unloading textures that he would be reloading shortly after, unless we are not caching them somehow - or unless they are taking up more memory than the allotted VRAM.

I believe that the texture memory manager does it in a last in first out logic, so unless we'd have queued so many textures already that we'd be unloading some that should be cached, I'm wondering why that'd be.

If you're developing on windows and can use a runtime debugger, you can certainly see what's the size of the loaded textures buffer, and what order they're in. Or we can create, for debugging purposes, a function to print the contents of that buffer (just the image paths, and loaded status, ordered), and we could call it before and after scrolling, to see what's happening.

I'm not sure that the example of scrolling up should unload image #6 (in your example) and then reload it unless we're doing something unexpected.

Koerty commented:
Ok, look good to me. Ready when you are @jrassa @pjft 

This hit kinda hard on fps at highest scroll speeds so I'm going to do many optimizations/refactoring in next PR before we enable the grid.

pjft commented:
Can you share what angles you attempted, just so we can think if there are any other things that come to mind?

Thank you!

Koerty commented:
@pjft Here is what I think of the current situation, I hope I didn't go too far when trying to understand your question.

So far I tried a couple of things, mostly toying with the event orders (by that I mean changing when tiles are build, when the dynamic loader make its changes, etc ...), I tried to add a dirty "hack", to make sure the tiles are not destroyed and then recreated when going in/out a folder, and a couple of other things I don't even remember atm.

I think if we want to properly address this issue, then we should take the problem at its root : the grid code, and the views code in general is way too much underoptimized, some events are triggered multiple times, some are even triggered in the wrong order.
For example : 
- I noticed the function onCursorChanged inherited from the IList, is called multiple times, for each time the cursor changed position, not sure why, it need further investigation.
- The tiles are builded when we first call the "render" function and "mEntriesDirty" is set to true, because that's the only point where we are sure the entries, the theme, and everything else were loaded right. Actually I'm not even sure of that since I have 2 systems on 6 that have already their "render" function called at ES start while still in system view.
- I noticed some similar problems at lower level, for example with the ImageComponent, if we call setImage and then setSize, we are basically setting the Texture, which trigger an automatic resize internally, then resize it to the final size. So at the grid level, if I'm not mistaken, we resize 2 times images, multiplied by the number of tiles, multipled by 3 since the "onCursorChanged" function is called 3 times so it may trigger 3 times the updateTiles function (not sure for this last one as I think I added another dirty fix to only update images if a texture were loaded or unloaded).

I think all of this should be addressed in a separate PR (or multiple one if changes are too big). First I would need to take a step back,, list all the events that happen (entries loaded, cursor changed, ...), when they should happen and what action they should trigger.

pjft commented:
Thanks for the detailed description, it's exactly what I was looking for. This way if we come up with alternatives we at least know if there's something you already tried or if it's something worth exploring.

I have found IList to be one of the weakest parts of the code in terms of error handling - when I was putting together the custom collections code the majority of the crashes I ran into came from there or from onXXXChanged events trying to access objects that had just been deleted, so I empathize.

Let me see if looking through the code sheds some light on this.

pjft commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -276,12 +308,188 @@ void ImageGridComponent<T>::onSizeChanged()
+{
+	if (mEntries[index].data.isTextureLoaded)
+	{
+		mEntries[index].data.texture = getDefaultTexture(mEntries[index].object);
```
In the sense that why do we even do this in the first place? Why not leave it with the actual texture? I wonder if this is not making the texture memory management a bit more constrained than it could.

pjft commented:
Hi @Koerty . Finally had the chance to stop and look through the code. I'd probably love to test this out somehow to have a better feel for what's happening, in the coming days/weeks when I have the time - is there a theme you'd recommend? Do share a link if you have it.

From reading the code on the View class, I'd strongly recommend testing this out extensively with collections, in particular adding and removing games to collections, empty collections, adding the first one, removing the last one (i.e. when there's only 1), etc.

I believe I recall doing a handful of fixes to the other views and IList in regards to segfaults that emerged from mishandling FileData objects, and I wonder which of those issues would still be present in this code and which ones would have been fixed.

I apologize but from reading the code I couldn't really spot anything extremely egregious that would suggest that the entering and exiting folders would eventually result in a segfault. It's just too much new code that's unknown to me that reading the code alone doesn't really suggest anything.

If you happen to pinpoint the crash to a line or method, I'm happy to look into it further. What I did in my case - since I only run the code on the Pi - is to use logs to determine/narrow down where things are crashing, the state of the objects, and from then work backwards. I know it's a pain, but if you don't have a runtime debugger on windows where you can replicate the crash easily, that'd be my recommendation.

Have a great weekend!

Koerty commented:
> Hi @Koerty . Finally had the chance to stop and look through the code. I'd probably love to test this out somehow to have a better feel for what's happening, in the coming days/weeks when I have the time - is there a theme you'd recommend? Do share a link if you have it.

Hi @pjft  Thanks for taking the time to go through all of this. You can find 3 "gridview-ready" themes on my GitHub profile in the pinned repository section.

> From reading the code on the View class, I'd strongly recommend testing this out extensively with collections, in particular adding and removing games to collections, empty collections, adding the first one, removing the last one (i.e. when there's only 1), etc.

Sure, but this is far from being ready to get merged. I will run more indepth tests like you suggested when the code will be more stable and clean.

> I apologize but from reading the code I couldn't really spot anything extremely egregious that would suggest that the entering and exiting folders would eventually result in a segfault. It's just too much new code that's unknown to me that reading the code alone doesn't really suggest anything.
>
> If you happen to pinpoint the crash to a line or method, I'm happy to look into it further. What I did in my case - since I only run the code on the Pi - is to use logs to determine/narrow down where things are crashing, the state of the objects, and from then work backwards. I know it's a pain, but if you don't have a runtime debugger on windows where you can replicate the crash easily, that'd be my recommendation.

I run Clion debugger on my Ubuntu, it help a lot usually, but in that case it doesn't provide much informations about the problem, or I were unable to understand them, idk.

The stack look like the following : 
- The grid is re-populated
- A call to the render function trigger the tile rebuild (`buildTiles();` on line 189)
- The buildTiles function clear the tiles before rebuilding them (`mTiles.clear();` on line 497)
- This delete the GridTileComponent
- Which delete the ImageGridComponent
- Which delete the TextureResource

At this last step ES crash, because the `std::shared_ptr<TextureResource>` expired at this point.

pjft commented:
Thanks, these help. If you can add some more line details there it helps even further, but just to confirm, it crashes on "deletion" of a TextureResource object?

pjft commented:
So, this is a bit complicated, but I do have a few questions:

`GridGameListView` has an `ImageGridComponent`, which in turn is an `IList<ImageGridData, T>`.

`ImageGridData` consists of:
```
struct ImageGridData
{
	bool isTextureLoaded;
	std::string texturePath;
	*std::shared_ptr<TextureResource>** texture;*
};
```
but then it also has
```
std::vector< std::shared_ptr<GridTileComponent> > mTiles;
```

`GridTileComponent` in turn has a
```
std::shared_ptr<ImageComponent> mImage;
```

which in turn has itself a 
```
std::shared_ptr<TextureResource> mTexture;
```

so my question is why are there two TextureResource shared pointers, and which one are we using for _what_?

I'm fairly sure that, between `populateList` which calls the clear() method, which in turn may call the destructor of each of the entries in that vector, so I suspect that between this back and forth of clearing and populating we're not properly managing memory for the images.

A very simple question:
- Why can't we have a fixed set of images in the grid (fixed as in "persistent objects", not in terms of size), and those images get updated with the relevant 

`mImage.setImage(file->getImagePath());`

when you scroll?

Maybe I'm misunderstanding the overall architecture here, but this whole thing of creating and deleting GuiComponents while navigating feels somewhat inefficient.

Also, from your description, something that may be happening is that, in `populateList`, the first thing it does is to call the `IList::clear()` method which will erase the `ImageGridComponent` contents and, in turn, a TextureResource for a specific path. If then, when you call `buildTiles`, the clear() method for the `mTiles` vector will also try to delete a TextureResource for the same path, maybe that's what's causing the issues?

My two cents here. Hopefully some things to explore or consider, but I'd certainly appreciate some further clarity on the architecture as I might certainly be completely off track here.

Have a good week.

Koerty commented:
> A very simple question:
>
> - Why can't we have a fixed set of images in the grid (fixed as in "persistent objects", not in terms of size), and those images get updated with the relevant
>
> mImage.setImage(file->getImagePath());
>
> when you scroll?

That's exactly what I proposed shortly after you talked me about the new texture management stuff, but since @jrassa and you wasn't so satisfied about it, and wanted to keep it close to what I firstly proposed in this PR, I gave up on it. We just need a bigger grid with more tiles that are not rendered (with clipRect) so we can load textures some rows ahead.

> Maybe I'm misunderstanding the overall architecture here, but this whole thing of creating and deleting GuiComponents while navigating feels somewhat inefficient.

Actually, this should only happen in the ImageGridComponent constructor or when the theme changed. But as I said some comments ago I need to take some step back to see what action happen and when they happen because until now I didn't find any other solution than build (or clear and rebuild) tiles at this point of the code. For example I tried to do it like I said (in constructor and on theme changed) but some events are triggered in the wrong order (like the onCursorChanged triggered at ES start I talked about in the other PR) which resulted in unexpected behavior (onCursorChanged triggered some actions depending on the default grid size, not the custom grid size defined by the theme).

> so my question is why are there two TextureResource shared pointers, and which one are we using for what?

Maybe I misunderstood the "shared pointer" thing, the point is for the tiles to re-use the pointers managed by the ImageGridComponent.

Generally speaking, you made fair points, and perfectly understood some of the ImageGridComponent design issues. As I said, I need to take a step back, see how we can re-architecture the ImageGridCompoent code and its parent the IList code to address some of this issues and create a cleaner base for further developments, where any addition or modification don't end up in this debugging hell where we can only understand what's happening by setting a few breakpoints and running a debugger.

Koerty commented:
@pjft Ok I got half an answer.

The crash is linked to an unexpected behavior due to a call of the "onCursorChanged" method, which happen when the IList.clear is called, before we repopulate the list.

If I comment the content of onCursorChanged, (the call to updateTargetRange and updateTiles) then it don't crash when entering/leaving a folder multiple times.

jrassa commented:
>> A very simple question:
>>
>> * Why can't we have a fixed set of images in the grid (fixed as in "persistent objects", not in terms of size), and those images get updated with the relevant
>> mImage.setImage(file->getImagePath());
>> 
>> when you scroll?
> 
> That's exactly what I proposed shortly after you talked me about the new texture management stuff, but since @jrassa and you wasn't so satisfied about it, and wanted to keep it close to what I firstly proposed in this PR, I gave up on it. We just need a bigger grid with more tiles that are not rendered (with clipRect) so we can load textures some rows ahead.

@Koerty I just want to say that for me, it wasn't that I wasn't satisfied with it, I even stated that it may be the best approach.  I was just concerned that it was maybe too big of a change from what you had in place and I didn't want to discourage you by telling you to throw it all out.  I was trying to find some middle ground.

Koerty commented:
@jrassa Ok it look good to me, I didn't encounter any major bug.

Beside that, it add so much complexity to the getStartPosition() function and to the updateTiles() function, I need to take a look at how make the code simpler, or at least clearer.

To test it out easily, you can revert the last commit (the one that add the clipRect), and build a theme with really small grid like this so the extra tiles used for buffering show on screen : 
```xml
<imagegrid name="gamegrid">
	<pos>0.047 0.3</pos>
	<size>.3 .3</size>
	<margin>0.03 0.03</margin>
</imagegrid>

<gridtile name="default">
	<size>0.05 0.06</size>
</gridtile>
```

Note that with the introduction of this extra rows and the clipRect, the last row can be partially shown, depending on the grid size and tile size. Right now it is a bug because the last displayed row is a row used for buffering images, but I actually like this bug so I'm considering fixing it but keep this behavior, i.e. rounding up the grid dimension y when I calculate the grid dimensions instead of rounding it down.

Koerty commented:
Rebased to resolve merge conflicts.

Koerty commented:
Fixed 2 bugs introduced by #428 : 
- a conflict between 428 and this PR which resulted in a crash of ES
- a bug overlooked in 428 where the tiles were not build if the theme have no "imagegrid" element

Koerty commented:
This PR was growing too big so I moved the 2 bug fixes mentioned in my last comment and the code related to the ClipRect to a new pull request : #430

@jrassa Let's focus on 430 first and then we will come back to this one.

Koerty commented:
Ok this look good to me.

The only issue I found is 80Mb vRAM limit is a bit short for the grid, I recommend setting it to at least 90~100 (of course this depend on a lot of external parameters like the scraped images size, the theme used, and so on ...)

pjft commented:
I imagine that that is a fair assessment. I don't oppose changing the default to 90 if it works well.

I reduced it from 100 to 80 because 100 would sometimes still result in issues, but with the amount of images being used in GridView it's certainly to be expected. 

How's the performance now, in your assessment? In specific, moving down and up - did the change in the loading loop make a difference? What worked and what didn't?

Thanks.

Koerty commented:
@pjft The change in the loading loop make nearly no difference now that I got back to TextureRessource pointers instead of strings. There is no more of this freezes when scrolling up even without it. It just look a little bit cleaner when it load images "backward" when scrolling up.

I actually thought about removing it, and make it always load images from top left to bottom right, as it complexify a code which is already complex enough but I just cleaned the code instead, and added some comments.

The performance is good enough right now, there is no big fps drop anymore, the grid will still need some optimizations in the future but this is out of the scope for now, the goal is to deliver this PR along with 419 and 433, so we have a first grid view "complete" implementation.

pjft commented:
Fantastic news, great to hear that the main issues were solved! Thanks for going through this. Looking forward to having this being tested by the forums and the wider community. 

I'd agree with keeping it as it is, if the performance is improved. Every bit counts on the Pi.

Thanks!

jrassa commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -344,26 +350,69 @@ void ImageGridComponent<T>::updateTiles()
-	int img = getStartPosition();
-
-	for(int ti = 0; ti < mTiles.size(); ti++)
+	// Stop updating the tiles at highest scroll speed
```
The other thing I noticed is that when you get to the fast scroll, the grid completely disappears while the letter overlay is rendered.  I understand we don't want to load the actually images, but could we just render the tiles with the default texture?

Koerty replied:
We can, it's a design choice I made. I didn't like how it looked with default texture so I made it completely fade out like metadata do. But I'm still open to discussion if you find that change necessary.

jrassa replied:
For now we can just leave it as is and then get some feedback from the wider user base once it is enabled.

Koerty commented:
> First is that the selection is being rendered on the truncated row.

This is an issue and I need to investigate why it happen. The selected tile should never end in the truncated row.

> Second, the truncated row is only rendered when a tile in it is selected.

I think the other tiles are either behind the bottom bar and help prompt, or out of the viewable area (the clipRect). The clipRect is not applied to the selected tile, as it can grow a bit out of the grid area.

> Third is that the amount of tile being partially rendered appears to be much larger then I would expect, based on the configured size.

I answered this with the answer to your second question. The clipRect is not applied to the selected tile, that's why it appear much bigger (in fact it should totally appear if the bottom part wasn't offscreen).

Of course this is just what is popping in my mind when reading about your problems. I need to investigate this issues and try it out to see if can come up with more informations or bug fixes.

jrassa commented:
>> Second, the truncated row is only rendered when a tile in it is selected.
> 
> I think the other tiles are either behind the bottom bar and help prompt, or out of the viewable area (the clipRect). The clipRect is not applied to the selected tile, as it can grow a bit out of the grid area.
> 
>> Third is that the amount of tile being partially rendered appears to be much larger then I would expect, based on the configured size.
> 
> I answered this with the answer to your second question. The clipRect is not applied to the selected tile, that's why it appear much bigger (in fact it should totally appear if the bottom part wasn't offscreen).
> 
> Of course this is just what is popping in my mind when reading about your problems. I need to investigate this issues and try it out to see if I can come up with more informations or bug fixes.

I missed that the selected tile is rendered outside the clipRect.  That makes sense.  I retested with a bigger size and verified that the truncated row rendered properly at all times.

Koerty commented:
@jrassa It still doesn't explain why the selected tile was in the truncated row. I can't seem to reproduce this bug. If you find more information about it, I'm interested.

>  I rechecked the math on the grid size and the optimal size based on the configured tile size and margin would be 0.64 vs 0.65.

Another thing, you should be able to set 0.64 or 0.65 or 0.66, it doesn't matter, because you have a 0.02 margin height so between 0.64 and 0.66 is our margin being displayed. The 4th row should only show up with value strictly superior to 0.66.

Koerty commented:
Ok I successfully reproduced the bug, it only happen when you have a gamelist containing between 10 and 12 games (included), with a grid size of 3x3 for example. I know where this come from and I'm working on a fix.

Koerty commented:
Ok I fixed it, it was a bugfix I added to #430 which weren't relevant anymore. So I ... fixed the now broken fix.

If you want to know the full story : 
This extra "if" were necessary before because the following were like this : "if we are close to the beggining; if we are close to the end;" and in this PR it got changed to "if we are close to the beggining **else** if we are close to the end". So this "if" were necessary because otherwise when we had just enough or less games than tiles on screen, both "if" got triggered and because of the 2nd one ("if we are close to the end") it was displaying the games starting at 2nd tile row, or skipped the first 3 games I'm not sure to remember correctly.

Btw, sorry for spamming, I just wanted to show you I'm giving much importance to your feedback after I half-turned down your change request on the behavior of the grid when scrolling at highest speed.

jrassa commented:
Sorry for the delay.  I just tested the latest and it is working, but I think it introduced another bug.  Same setup with a 3x3 grid.  I have a system that only has 5 games.  When scrolling on this system the grid is initially rendered with 3 tiles in the first row, 2 tiles in the second row and none in the 3rd (as expected).  But when I move the cursor to the 2 second row, the tiles shift down such that there are no tiles in the first row.

Koerty commented:
@jrassa Okay I completely fixed this, did a lot of testing and squashed.

This look to get merged to me unless you have anything else to say.
